namespace/id.go

```diff
@@ -0,0 +1,24 @@
+package namespace
+
+import "bytes"
+
+type ID []byte
```

Comment:
Is there a way to make this a fixed-size byte array? I know Go doesn't have generics, but it's a little opaque to have the ID length only set at runtime when it's a parameter that can be known at compile time. Maybe using constants?

Reply:
I don't think there is an elegant way to achieve this in golang but still allow the user to define the size they like. An alternative would be to make the ID length/size part of the "prefixed data" type (I miss associated types ...), which then could return an ID (of fixed length) depending on the concrete implementation (similar to https://golang.org/pkg/crypto/sha256/#Sum256).

Reply:
Ugh 🤷. My concern is that variable-sized arrays might serialize differently than fixed-size arrays, depending on which canonical serialization scheme is used. And this library should be parametrizable with different serialization schemes.

Reply:
> My concern is that variable-sized arrays might serialize differently than fixed-size arrays, depending on which canonical serialization scheme is used.

The length will be fixed for any tree even with the current implementation. The type doesn't reflect this to be flexible. Similar to the `Sum` method in https://golang.org/pkg/hash/#Hash which also always return the same size for a concrete implementation of a hash function.

> And this library should be parametrizable with different serialization schemes.

Hmm, good point. I'll think about how to abstract this cleanly. Is it OK, if I document this better for this PR and think about this in future iterations?

Reply:
> And this library should be parametrizable with different serialization schemes.

Thinking about this further: I don't think the NMT should care about serialization. The whole API and the implementation would be simpler if the caller would need to pass in the data already properly serialized (as bytes / simple type aliases).

Also, I think going back to my original suggestion to have `Push()` take in the pair of namespace and data separately instead of some wrapping type makes more sense.

Reply:
So one problem is that hash digests are also a data type and also need a canonical serialization. This can't be done at the input level because hashing is an internal operation to the Merkle tree.

Reply:
By the hash digests you mean the resulting root, right? These are also just bytes. Can't the consumer serialize these as he wishes? Isn't this a bit like saying the sha256 `hash.Hash`  additionally should define a canonical serialization format for its output :thinking: E.g. the SMT doesn't define how to pass the resulting root bytes over the wire. Why would it be different here?

Can you clarify the expectations the point you are trying to make a bit further?